File: 3-leetcode/01-easy/cpp/590-n-ary-tree-postorder-traversal.cpp
#include <bits/stdc++.h>

using namespace std;

class Node
{
public:
    int val;
    vector<Node *> children;

    Node() {}

    Node(int _val)
    {
        val = _val;
    }

    Node(int _val, vector<Node *> _children)
    {
        val = _val;
        children = _children;
    }
};

class Solution
{
public:
    vector<int> postorder(Node *root)
    {
        traverse(root);

        return result;
    }

private:
    vector<int> result;

    void traverse(Node *node)
    {
        if (!node)
        {
            return;
        }

        for (Node *child : node->children)
        {
            traverse(child);
        }

        result.push_back(node->val);
    }
};

File: 3-leetcode/01-easy/js/590-n-ary-tree-postorder-traversal.ts
class _Node {
  val: number;
  children: _Node[];
  constructor(val?: number) {
    this.val = val === undefined ? 0 : val;
    this.children = [];
  }
}

const postorder = (root: _Node | null): number[] => {
  const result = [];

  traverse(root, result);

  return result;
};

function traverse(node: _Node, result: number[]) {
  if (node == null) {
    return;
  }

  for (let child of node.children) {
    traverse(child, result);
  }

  result.push(node.val);
}
